File: app/src/main/java/com/kickstarter/viewmodels/ProjectSocialViewModel.java
package com.kickstarter.viewmodels;

import android.support.annotation.NonNull;

import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.Environment;
import com.kickstarter.models.Project;
import com.kickstarter.ui.IntentKey;
import com.kickstarter.ui.activities.ProjectSocialActivity;
import com.kickstarter.viewmodels.outputs.ProjectSocialViewModelOutputs;

import rx.Observable;
import rx.subjects.BehaviorSubject;

public final class ProjectSocialViewModel extends ActivityViewModel<ProjectSocialActivity> implements ProjectSocialViewModelOutputs {

  private final BehaviorSubject<Project> project = BehaviorSubject.create();
  @Override
  public Observable<Project> project() {
    return project;
  }

  public final ProjectSocialViewModelOutputs outputs = this;

  public ProjectSocialViewModel(final @NonNull Environment environment) {
    super(environment);

    intent()
      .map(i -> i.getParcelableExtra(IntentKey.PROJECT))
      .ofType(Project.class)
      .compose(bindToLifecycle())
      .subscribe(project);
  }
}

File: app/src/main/java/com/kickstarter/viewmodels/ProjectSocialViewModel.kt
package com.kickstarter.viewmodels

import android.content.Intent
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.kickstarter.libs.Environment
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.models.Project
import com.kickstarter.ui.IntentKey
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject

interface ProjectSocialViewModel {
    interface Outputs {
        fun project(): Observable<Project>
    }

    class ProjectSocialViewModel(private val environment: Environment, private val intent: Intent? = null) :
        ViewModel(), Outputs {

        private val project = BehaviorSubject.create<Project>()
        private val disposables = CompositeDisposable()

        private fun intent() = intent?.let { Observable.just(it) } ?: Observable.empty()

        val outputs: Outputs = this

        init {
            intent()
                .map<Any> { it.getParcelableExtra(IntentKey.PROJECT) }
                .ofType(Project::class.java)
                .subscribe {
                    project.onNext(it)
                }
                .addToDisposable(disposables)
        }

        override fun project(): Observable<Project> = project

        override fun onCleared() {
            disposables.clear()
            super.onCleared()
        }
    }

    class Factory(private val environment: Environment, private val intent: Intent? = null) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return ProjectSocialViewModel(environment, intent) as T
        }
    }
}
